src/amltk/scheduling/plugins/emissions_tracker_plugin.py

```diff
@@ -0,0 +1,118 @@
+
+from amltk.scheduling.plugins.plugin import Plugin
+from amltk.scheduling.task import Task
+
+P = TypeVar("P")
+R = TypeVar("R")
+
+
+class _EmissionsTrackerWrapper:
+    """A wrapper around codecarbon package to measure emissions."""
+
+    def __init__(
+            self,
+            fn: Callable[[P], R],
+            task: Task,
+            *codecarbon_args: Any,
+            **codecarbon_kwargs: Any,
+    ):
+        """Initialize the wrapper.
+
+        Args:
+            fn: The function to wrap.
+            task: The task the function is being wrapped for.
+            *codecarbon_args: arguments to pass to
+                [`codecarbon.EmissionsTracker`][codecarbon.EmissionsTracker].
+            **codecarbon_kwargs: keyword arguments to pass to
+                [`codecarbon.EmissionsTracker`][codecarbon.EmissionsTracker].
+        """
+        super().__init__()
+        self.fn = fn
+        self.task = task
+        self.codecarbon_args = codecarbon_args
+        self.codecarbon_kwargs = codecarbon_kwargs
+
+    def __call__(self, *args: any, **kwargs: any) -> R:
+        with EmissionsTracker(*self.codecarbon_args, **self.codecarbon_kwargs) as tracker:
+            result = self.fn(*args, **kwargs)
+            return result
+
+
+class EmissionsTrackerPlugin(Plugin):
+    """A plugin that tracks carbon emissions using codecarbon library."""
+
+    name: ClassVar = "emissions-tracker"
+    """The name of the plugin."""
+
+    """
+    Usage Example:
+
+    ```python
+    from concurrent.futures import ThreadPoolExecutor
+    from amltk.scheduling import Scheduler
+    from amltk.scheduling.plugins.emissions_tracker_plugin import EmissionsTrackerPlugin
+
+    def some_function(x: int) -> int:
+        return x * 2
+
+    executor = ThreadPoolExecutor(max_workers=1)
+
+    # Create a Scheduler instance with the executor
+    scheduler = Scheduler(executor=executor)
+
+    # Create a task with the emissions tracker plugin
+    task = scheduler.task(some_function, plugins=[
+        EmissionsTrackerPlugin(log_level="info", save_to_file=False) # pass any codecarbon args here
+    ])
+
+    @scheduler.on_start
+    def on_start():
+        task.submit(5) # submit any args here
+
+    @task.on_submitted
+    def on_submitted(future, *args, **kwargs):
+        print(f"Task was submitted", future, args, kwargs) 
+
+    @task.on_done
+    def on_done(future):
+        print("Task done: ", future.result()) # result is the return value of the function
+
+    scheduler.run()    
+    ```  
+    """
+
+    def __init__(self, *args: Any, **kwargs: Any):
+        super().__init__()
+        self.task: Task | None = None
+        self.codecarbon_args = args
+        self.codecarbon_kwargs = kwargs
+
+    def attach_task(self, task: Task) -> None:
+        """Attach the plugin to a task."""
+        self.task = task
+
+    def pre_submit(
+            self,
+            fn: Callable[[P], R],
+            *args: any,
+            **kwargs: any,
```

Comment:
One of the failing github actions is the linter, did you run `ruff` on the code or `pre-commit`?

Reply:
Oh, you're right, thank you. Do you know how to add the linter to the IDE? So it would automatically highlight the lines that are too long etc.

Reply:
@LennartPurucker 

As far as I'm aware, there's a plugin for it for pycharm and you can also revamp the format hotkey to use ruff instead

Reply:
Heyho, see this https://plugins.jetbrains.com/plugin/20574-ruff. I had to play around a bit to make it work though.

Reply:
Thank you! That's very helpful.